### PR TITLE
Send errors metrics for 5xx response from API Gateway, Lambda Function URL, or ALB

### PR DESCRIPTION
**Pull request by kimi-p**
### What does this PR do?
- Send `aws.lambda.enhanced.errors` when the Lambda invocation is from AWS API Gateway, AWS Lambda Function URL, or ALB. 
- Sets `span.error=1`.
- Reset `self.response` to `None` in `_before` so that the response from last invocation wouldn't carry over into the current invocation.

### Motivation
When using a web framework, and something went wrong with a request, instead of raising an exception, the web framework typically returns a `5xx` response back to the client. From user's perspective, invocations returning a `5xx` response code (aka, the value of the statusCode field from the response is `5xx`) should be reported as errors in Datadog.

### Testing Guidelines

image

image

### Additional Notes
To avoid sending duplicated `aws.lambda.enhanced.errors` metric, `self.already_submitted_errors_metric_before` is added. 

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: datadog_lambda/wrapper.py [hunk 1]
```diff
@@ -190,6 +192,9 @@ def _after(self, event, context):
             status_code = extract_http_status_code_tag(self.trigger_tags, self.response)
             if status_code:
                 self.trigger_tags["http.status_code"] = status_code
+                if not self.already_submitted_errors_metric_before:
```
[the remark this answers is not in the document]
Comment (tianchu) [reply datadog_lambda/wrapper.py:195]: BTW, in your current implementation `already_submitted_errors_metric_before` never resets to `False` once set `True`. `__init__` only gets called once on cold start. You need to reset it in `_before` (or in `_call` before the `try` block)

File: datadog_lambda/wrapper.py [hunk 2]
```diff
@@ -190,6 +192,9 @@ def _after(self, event, context):
...
+                    if len(status_code) == 3 and status_code.startswith("5"):
+                        submit_errors_metric(context)
```
Comment (tianchu) [inline datadog_lambda/wrapper.py:197]: You also need to mark `self.span` as error, otherwise the trace and metric won't match.

File: datadog_lambda/wrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimi-p) [reply datadog_lambda/wrapper.py:195]: Nice catch. `self.response` is already initialized to `None` here: https://github.com/DataDog/datadog-lambda-python/blob/de5aba059c2c87d83e10622144481bb80f8fe14d/datadog_lambda/wrapper.py#L112-L112 .

And when the exception happens, `self.response` will still be `None`. So, we actually do not need the flag (`self. already_submitted_errors_metric_before `) to avoid sending errors metrics twice since the `statusCode` will be `None`.

File: datadog_lambda/wrapper.py (hunk 2 above)
Comment (kimi-p) [reply datadog_lambda/wrapper.py]: Please let me know if I miss anything to set `self.span` as error. Thanks!
https://github.com/DataDog/datadog-lambda-python/pull/229/commits/ad09cc90874f0d911637a939cd295968ff308ea8#diff-5efc779a7e3c73ebd812eb62ff5dce68db1e0ed9ef9765bb076638733e8c5ae6R195-R196

File: datadog_lambda/wrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tianchu) [reply datadog_lambda/wrapper.py]: Yes, it's initialized to None, but it doesn't get reset to None between invocations. E.g., you have a good invocation, which would set self.response, then you have a failed invocation, self.response won't reset to None. extract_http_status_code_tag would extract the status code from the last good invocation instead of from the current failed invocation.

Even putting your use case aside, the state of self.response probably should be reset to None on before every invocation, keeping the invocation response value from the previous invocation in the current invocation is meaningless and can only lead to bugs.

[the remark this answers is not in the document]
Comment (tianchu) [reply datadog_lambda/wrapper.py]: Maybe one context you missed here is that Lambda isn't truly stateless, a sandbox/container is reused for many invocations until it gets recycled by AWS. https://pfisterer.dev/posts/aws-lambda-container-reuse might be a good read?

File: datadog_lambda/wrapper.py [hunk 3]
```diff
@@ -190,6 +190,10 @@ def _after(self, event, context):
             status_code = extract_http_status_code_tag(self.trigger_tags, self.response)
             if status_code:
                 self.trigger_tags["http.status_code"] = status_code
+                if len(status_code) == 3 and status_code.startswith("5"):
+                    submit_errors_metric(context)
+                    if self.span:
+                        self.span.set_traceback()
```
Comment (tianchu) [inline datadog_lambda/wrapper.py:196]: I'm not sure you have a valid stacktrace in this case, since there isn't a real python exception being thrown. I suspect you need to directly set the error message and type fields https://github.com/DataDog/dd-trace-py/blob/fb8dfa2f33fff37d21df9728d8386c0260df9744/ddtrace/contrib/grpc/server_interceptor.py#L41-L42

We should set both fields to something meaningful that can help users understand the problem when seeing it. For error.type, we can use `StatusCode 5xx`, for error.msg we can probably say `Lambda invocation returns status code 500` (we can probably spell out the actual status code in the message instead of 5xx).

@brettlangdon we are trying to mark Lambda invocations returning statusCode 5xx as errors in trace, is what I mentioned above the best way to do it?

Comment (brettlangdon) [reply datadog_lambda/wrapper.py:196]: This is what we do in the tracer: https://github.com/DataDog/dd-trace-py/blob/3b91b0da8/ddtrace/contrib/trace_utils.py#L274-L282

tl;dr; you just need to do `self.span.error = 1`

File: datadog_lambda/wrapper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimi-p) [reply datadog_lambda/wrapper.py]: Thanks for sharing the post! Just want to be sure (since the post mentioned only about shared folder -- `It's possible that a container is reused, which causes the /tmp folder to be shared between invocations.`), does the singleton wrapper got reused between multiple invocations?
```datadog_lambda_wrapper = _LambdaDecorator```

I plan to reset `self.response` at the beginning of `_before()` anyway. Thanks for calling this out!

[the remark this answers is not in the document]
Comment (tianchu) [reply datadog_lambda/wrapper.py]: yes, a new container starts up on first invocation, your python app get loaded into memory, that's called cold start, then from there, the container will be reused for subsequent invocations, until AWS recycles it, or AWS needs to launch more containers to handle more requests (one lambda container can only handle one request/invocation at a time)

File: datadog_lambda/wrapper.py [hunk 4]
```diff
@@ -190,6 +194,18 @@ def _after(self, event, context):
...
+                        self.span.error = 1
+                        self.span.set_tags(
+                            {
+                                ERROR_TYPE: "5xx Server Errors",
+                                ERROR_MSG: SERVER_ERRORS_STATUS_CODES.get(
+                                    status_code, "5xx Server Errors"),
```
Comment (kimi-p) [inline datadog_lambda/wrapper.py:206]: I wonder if we should set `error.trace` here as some hard-coded string since the stack trace of `span.set_traceback()` is not helpful and may be confusing.

image

File: datadog_lambda/wrapper.py [hunk 5]
```diff
@@ -150,6 +153,7 @@ def __call__(self, event, context, **kwargs):
             self._after(event, context)
 
     def _before(self, event, context):
+        self.response = None
```
Comment (tianchu) [inline datadog_lambda/wrapper.py:156]: let's put this line inside the `try:...` block, so we avoid datadog code crashing customer application as much as we can....that is, datadog should fail quietly (only log something) but no interruption to customer application. I know this line is pretty safe, but it may "invite" future developers to add more lines outside the try block because they saw some lines outside...

File: datadog_lambda/wrapper.py [hunk 6]
```diff
@@ -190,6 +193,16 @@ def _after(self, event, context):
             status_code = extract_http_status_code_tag(self.trigger_tags, self.response)
             if status_code:
                 self.trigger_tags["http.status_code"] = status_code
+                if len(status_code) == 3 and status_code.startswith("5"):
+                    submit_errors_metric(context)
+                    if self.span:
+                        self.span.set_traceback()
+                        self.span.error = 1
+                        self.span.set_tags({
+                            ERROR_TYPE: "5xx Server Errors",
```
[the remark this answers is not in the document]
Comment (kimi-p) [reply datadog_lambda/wrapper.py]: Hey Brett,

I am new to the team so might not be 100% correct, but I think we want to mark 5xx responses as "Error" even though the program does not throw any exception. The intended result UI I think looks like the first graph. So that's why I'm setting the error.type and error.msg here (according the 2nd picture from the datadog doc).
image

image


> Do we also tag this span with the http.status_code ?
Yes, we do here right before the logic I added: https://github.com/DataDog/datadog-lambda-python/blob/d02855ad1833e6a3157908c5d8c00d6afa6a9e36/datadog_lambda/wrapper.py#L195-L195

[the remark this answers is not in the document]
Comment (brettlangdon) [reply datadog_lambda/wrapper.py]: Typically when we mark a span as an error because of the status code we don't set a traceback (e.g. don't we don't call `span.set_traceback()` because there isn't an exception/traceback associated with the "failure").

it is ok to only set `span.error = 1`, as long as we have the `http.status_code` tag also set on the span (thanks for clarifying, I couldn't tell if `trigger_tags` get added to the span or not) then the UI knows what to do with it.

But this is probably more of a product decision, do we want to add a traceback for these specific 500 responses? Do they provide value for customers?

[the remark this answers is not in the document]
Comment (brettlangdon) [reply datadog_lambda/wrapper.py]: (also 👋🏻 welcome!!)

[the remark this answers is not in the document]
Comment (kimi-p) [reply datadog_lambda/wrapper.py]: Regarding `trigger_tags`, can't be 100% sure but I think we did set it here: https://github.com/DataDog/datadog-lambda-python/blob/a7cb6ef0cd61953d170c50cdbbb78d7a0958c615/datadog_lambda/wrapper.py#L173-L181

I will check with the team about the traceback. Thanks for the comment.

[the remark this answers is not in the document]
Comment (kimi-p) [reply datadog_lambda/wrapper.py]: Update: we're only setting `self.error=1` and not setting the `error.type` and `error.msg` now. Frontend will handle like what APM does. The PR description is updated to reflect the change and screenshots are also updated.

File: datadog_lambda/constants.py [hunk 7]
```diff
@@ -42,3 +42,13 @@ class XrayDaemon(object):
     XRAY_TRACE_ID_HEADER_NAME = "_X_AMZN_TRACE_ID"
     XRAY_DAEMON_ADDRESS = "AWS_XRAY_DAEMON_ADDRESS"
     FUNCTION_NAME_HEADER_NAME = "AWS_LAMBDA_FUNCTION_NAME"
+
+
+SERVER_ERRORS_STATUS_CODES = {
```
Comment (tianchu) [inline datadog_lambda/constants.py:47]: you probably don't need them any longer.

**Review by tianchu — APPROVED** [review]
💯
